### PR TITLE
feat: [PL-25452]: Enhance SCM Error Handling with more details

### PR DESCRIPTION
**Pull request by mohitg0795**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/34131)

### CODE REVIEW

**Review by ankit-harness — DISMISSED** [review]
:lgtm:

> Reviewed 61 of 61 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34131#-:-N46XzMq6aT_0ClCw170:bnfp4nl)* status: 61 file(s) to review. 1 of 2 approvals obtained.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/ScmApiErrorHandlingHelper.java [hunk 1]
```diff
@@ -25,10 +26,20 @@
 @Slf4j
 @OwnedBy(PL)
 public class ScmApiErrorHandlingHelper {
+  public void processAndThrowError(ScmApis scmAPI, ConnectorType connectorType, String repoUrl, int statusCode,
+      String errorMessage, ErrorMetadata errorMetadata) {
+    if (errorMetadata == null) {
+      errorMetadata = ErrorMetadata.builder().build();
+    }
+
```
Comment (bhavya181) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/ScmApiErrorHandlingHelper.java:34]: Why is this method required?

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateBranchScmApiErrorHandler.java [hunk 2]
```diff
@@ -15,44 +15,54 @@
...
   public static final String CREATE_BRANCH_FAILED_MESSAGE =
-      "The requested branch could not be created on Bitbucket Server. ";
+      "The requested branch<BRANCH> could not be created on Bitbucket Server. ";
   public static final String CREATE_BRANCH_UNPROCESSABLE_ENTITY_ERROR_HINT =
-      "Please check that the requested Bitbucket branch name is valid and does not already exist.";
+      "Please check that the requested Bitbucket branch name<BRANCH> is valid and does not already exist.";
   public static final String CREATE_BRANCH_UNPROCESSABLE_ENTITY_ERROR_EXPLANATION = "Possible reasons can be:\n"
-      + "1. A branch with given name already exists in the remote Bitbucket repository.\n"
-      + "2. The given branch name is invalid.";
+      + "1. A branch with given name<BRANCH> already exists in the remote Bitbucket repository<REPO>.\n"
```
Comment (bhavya181) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateBranchScmApiErrorHandler.java:32]: newbranch name

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateBranchScmApiErrorHandler.java [hunk 3]
```diff
@@ -15,44 +15,54 @@
...
+      + "2. The given branch name<BRANCH> is invalid.";
```
Comment (bhavya181) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateBranchScmApiErrorHandler.java:33]: new branch name

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetDefaultBranchScmApiErrorHandler.java [hunk 4]
```diff
@@ -25,16 +27,20 @@ public class BitbucketServerGetDefaultBranchScmApiErrorHandler implements ScmApi
   public static final String GET_DEFAULT_BRANCH_FAILED_MESSAGE = "Fetching default branch from Bitbucket failed. ";
```
Comment (bhavya181) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetDefaultBranchScmApiErrorHandler.java:27]: put branch name

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/ScmApiErrorHandlingHelper.java` line 34 at r1](https://reviewable.io/reviews/harness/harness-core/34131#-N46SsBiG1kySfN_h3JG-r1-34:-N48vDKI0OQwI2JCH_0O:b-9ik5r6) ([raw file](https://github.com/harness/harness-core/blob/eda37cdca298003ac645e81e0257c4b483423fe3/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/ScmApiErrorHandlingHelper.java#L34)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

Why is this method required?
</blockquote></details>

How will errorMetadata come as input? Currently, we don't have errorMetadata as input

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketcloud/BitbucketGetDefaultBranchScmApiErrorHandler.java` line 27 at r1](https://reviewable.io/reviews/harness/harness-core/34131#-N46SsBiG1kySfN_h3JM-r1-27:-N48v_K28ETXLW0RUnWQ:b-58v9ka) ([raw file](https://github.com/harness/harness-core/blob/eda37cdca298003ac645e81e0257c4b483423fe3/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketcloud/BitbucketGetDefaultBranchScmApiErrorHandler.java#L27)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

Have we not messages in all apis?
</blockquote></details>

Didn't get you

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetDefaultBranchScmApiErrorHandler.java` line 27 at r1](https://reviewable.io/reviews/harness/harness-core/34131#-N46SsBjjp-2HbMuRj-G-r1-27:-N48vdQ14dhB-eL1KDd1:bj1t3a4) ([raw file](https://github.com/harness/harness-core/blob/eda37cdca298003ac645e81e0257c4b483423fe3/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerGetDefaultBranchScmApiErrorHandler.java#L27)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

put branch name
</blockquote></details>

here branch won't come. This is failure in getDefaultBranch API.

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketcloud/BitbucketCreateBranchScmApiErrorHandler.java` line 32 at r1](https://reviewable.io/reviews/harness/harness-core/34131#-N46SsBiG1kySfN_h3JJ-r1-32:-N4BdzIi9-r8LNYKAbyS:b-zfthoj) ([raw file](https://github.com/harness/harness-core/blob/eda37cdca298003ac645e81e0257c4b483423fe3/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketcloud/BitbucketCreateBranchScmApiErrorHandler.java#L32)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

NewBranchName field will be used here.
</blockquote></details>

Yes, correct. Also in the message above it, there also. Fixed at both places.

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketcloud/BitbucketCreateBranchScmApiErrorHandler.java` line 33 at r1](https://reviewable.io/reviews/harness/harness-core/34131#-N46SsBiG1kySfN_h3JJ-r1-33:-N4Be7YGDsCUFCMKZUAr:bii7s6r) ([raw file](https://github.com/harness/harness-core/blob/eda37cdca298003ac645e81e0257c4b483423fe3/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketcloud/BitbucketCreateBranchScmApiErrorHandler.java#L33)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

NewBranchName field will be used.
</blockquote></details>

correct, fixed.

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateBranchScmApiErrorHandler.java` line 32 at r1](https://reviewable.io/reviews/harness/harness-core/34131#-N46SsBjjp-2HbMuRj-D-r1-32:-N4BeLsZ8Cjk5YPrJdhc:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/eda37cdca298003ac645e81e0257c4b483423fe3/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateBranchScmApiErrorHandler.java#L32)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

newbranch name
</blockquote></details>

done

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateBranchScmApiErrorHandler.java` line 33 at r1](https://reviewable.io/reviews/harness/harness-core/34131#-N46SsBjjp-2HbMuRj-D-r1-33:-N4BeMbP0HuHH4QvMgdg:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/eda37cdca298003ac645e81e0257c4b483423fe3/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/bitbucketserver/BitbucketServerCreateBranchScmApiErrorHandler.java#L33)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

new branch name
</blockquote></details>

done

**Comment by mohitg0795** [comment]
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java` line 32 at r1](https://reviewable.io/reviews/harness/harness-core/34131#-N46SsBjjp-2HbMuRj-N-r1-32:-N4BeZX119bs6NQZdFAF:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/eda37cdca298003ac645e81e0257c4b483423fe3/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/handlers/github/GithubCreateBranchScmApiErrorHandler.java#L32)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

new branch name
</blockquote></details>

done

**Review by bhavya181 — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34131)* status: 61 file(s) to review. 1 of 2 approvals obtained. 

___
*[`136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/ScmApiErrorHandlingHelper.java` line 34 at r1](https://reviewable.io/reviews/harness/harness-core/34131#-N46SsBiG1kySfN_h3JG-r1-34:-N4W-ApDCh4KkyGZkfWs:b6bui0b) ([raw file](https://github.com/harness/harness-core/blob/eda37cdca298003ac645e81e0257c4b483423fe3/136-git-sync-manager/src/main/java/io/harness/gitsync/common/scmerrorhandling/ScmApiErrorHandlingHelper.java#L34)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

How will errorMetadata come as input? Currently, we don't have errorMetadata as input
</blockquote></details>

ya, basically I was talking about previous method, but we had a discussion over it, so, we can go ahead with two methods

**Review by kapil-harness — APPROVED** [review]
> Reviewed 44 of 61 files at r1, 17 of 17 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34131)* status: 17 file(s) to review.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 44 of 61 files at r1, 17 of 17 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34131)* status: :shipit: complete! ✓

**Pull request merged**